src/lib/ares_init.c

```diff
@@ -193,66 +195,76 @@ int ares_init_options(ares_channel *channelptr, struct ares_options *options,
                    ares_strerror(status)));
-    /* If we fail to apply user-specified options, fail the whole init process */
+    /* If we fail to apply user-specified options, fail the whole init process
+     */
     goto done;
   }
   status = init_by_environment(channel);
-  if (status != ARES_SUCCESS)
+  if (status != ARES_SUCCESS) {
     DEBUGF(fprintf(stderr, "Error: init_by_environment failed: %s\n",
                    ares_strerror(status)));
+  }
   if (status == ARES_SUCCESS) {
     status = init_by_resolv_conf(channel);
-    if (status != ARES_SUCCESS)
+    if (status != ARES_SUCCESS) {
       DEBUGF(fprintf(stderr, "Error: init_by_resolv_conf failed: %s\n",
                      ares_strerror(status)));
+    }
   }
 
   /*
    * No matter what failed or succeeded, seed defaults to provide
    * useful behavior for things that we missed.
    */
   status = init_by_defaults(channel);
-  if (status != ARES_SUCCESS)
+  if (status != ARES_SUCCESS) {
     DEBUGF(fprintf(stderr, "Error: init_by_defaults failed: %s\n",
                    ares_strerror(status)));
+  }
 
   /* Trim to one server if ARES_FLAG_PRIMARY is set. */
-  if ((channel->flags & ARES_FLAG_PRIMARY) && channel->nservers > 1)
+  if ((channel->flags & ARES_FLAG_PRIMARY) && channel->nservers > 1) {
     channel->nservers = 1;
+  }
 
   status = ares__init_servers_state(channel);
   if (status != ARES_SUCCESS) {
     goto done;
   }
 
 done:
-  if (status != ARES_SUCCESS)
-    {
-      /* Something failed; clean up memory we may have allocated. */
-      if (channel->servers) {
-        ares_free(channel->servers);
-      }
-      if (channel->ndomains > 0)
-        ares__strsplit_free(channel->domains, channel->ndomains);
-      if (channel->sortlist)
-        ares_free(channel->sortlist);
-      if(channel->lookups)
-        ares_free(channel->lookups);
-      if(channel->resolvconf_path)
-        ares_free(channel->resolvconf_path);
-      if(channel->hosts_path)
-        ares_free(channel->hosts_path);
-      if (channel->rand_state)
-        ares__destroy_rand_state(channel->rand_state);
-
-      ares__htable_stvp_destroy(channel->queries_by_qid);
-      ares__llist_destroy(channel->all_queries);
-      ares__slist_destroy(channel->queries_by_timeout);
-      ares__htable_asvp_destroy(channel->connnode_by_socket);
-      ares_free(channel);
-      return (int)status;
+  if (status != ARES_SUCCESS) {
+    /* Something failed; clean up memory we may have allocated. */
+    if (channel->servers) {
+      ares_free(channel->servers);
+    }
+    if (channel->ndomains > 0) {
+      ares__strsplit_free(channel->domains, channel->ndomains);
+    }
+    if (channel->sortlist) {
+      ares_free(channel->sortlist);
+    }
+    if (channel->lookups) {
+      ares_free(channel->lookups);
+    }
+    if (channel->resolvconf_path) {
+      ares_free(channel->resolvconf_path);
+    }
+    if (channel->hosts_path) {
+      ares_free(channel->hosts_path);
+    }
```

Comment:
ares_free() handles NULL by itself, maybe it is time to stop doing the superfluous NULL checks?